cli: display multiwallet balances in -getinfo

jonatack opened the pull request:
This PR is a client-side version of #18453, per review feedback there and [review club discussions](https://bitcoincore.reviews/18453#meeting-log). It updates `bitcoin-cli -getinfo` on the client side to display wallet name and balance for the loaded wallets when more than one is loaded (e.g. you are in "multiwallet mode") and `-rpcwallet=` is not passed; otherwise, behavior is unchanged.

before
```json
$ bitcoin-cli -getinfo -regtest
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "balance": 0.00001000,
  "relayfee": 0.00001000
}

```
after
```json
$ bitcoin-cli -getinfo -regtest
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "balances": {
    "": 0.00001000,
    "Encrypted": 0.00003500,
    "day-to-day": 0.00000120,
    "side project": 0.00000094
  }
}
```
-----

`Review club` discussion about this PR is here: https://bitcoincore.reviews/18453

This PR can be manually tested by building, creating/loading/unloading several wallets with `bitcoin-cli createwallet/loadwallet/unloadwallet` and running `bitcoin-cli -getinfo` and `bitcoin-cli -rpcwallet=<wallet-name> -getinfo`.

`wallet_multiwallet.py --usecli` provides regression test coverage on this change, along with `interface_bitcoin_cli.py` where this PR adds test coverage.

Credit to Wladimir J. van der Laan for the idea in https://github.com/bitcoin/bitcoin/issues/17314 and https://github.com/bitcoin/bitcoin/pull/18453#issuecomment-605431806.

jonatack committed:
cli: lift -rpcwallet logic up to CommandLineRPC()

to allow passing rpcwallet independently from the -rpcwallet user option, and to
move the logic to the top-level layer where most of the other option args are
handled.

jonatack committed:
test: add -getinfo multiwallet functional tests

and improve the existing -getinfo -rpcwallet tests.

jonatack commented:
Force-pushed after seeing a way to simplify and improve it.

promag commented:
Concept ACK

interface_bitcoin_cli.py  fails in no wallet job.

jonatack commented:
Thanks -- will look.

fanquake commented:
It seems this is in part reverting changes that were just merged in #18574 (split out of #18453), and refactoring to handle multiwallet.

I understand splitting up changes, and maybe it doesn't matter so much in this instance, but it's a bit of code/review churn if we're PR'ing and merging refactors only to essentially undo the changes and do something else in a different PR a few hours later.

jonatack commented:
I empathise. See https://github.com/bitcoin/bitcoin/pull/18453#issuecomment-611429161.  #18574 was the only change that seemed to have consensus, so it seemed best to split it out for merge and propose client-side code in this PR to test and compare with the server-side code in #18453. Sometimes having and running working code helps debates move forward.

jonatack committed:
cli: extract connection exception handler, -rpcwait logic

to ConnectAndCallRPC() to be callable for individual connections.

This is needed for RPCs that need to be called and handled sequentially, rather
than alone or in a batch.

For example, when fetching the balances for each loaded wallet, -getinfo will
call RPC listwallets, and then, depending on the result, RPC getbalances.

It may be somewhat helpful to review this commit with `git show -w`.

jonatack committed:
cli: create GetWalletBalances() to fetch multiwallet balances

jonatack committed:
rpc: drop unused JSONRPCProcessBatchReply size arg, refactor

jonatack commented:
Updated with the following:
- ensure it works when built without the wallet
- extract the connection try/wait/failure logic to be callable for each connection, so we can use this for calling `listwallets` and `getbalances` to be more robust
- drop an unused arg in `JSONRPCProcessBatchReply` as requested in https://github.com/bitcoin/bitcoin/pull/18574#discussion_r406526973

jnewbery commented on src/bitcoin-cli.cpp:
```diff
@@ -485,56 +500,58 @@ static int CommandLineRPC(int argc, char *argv[])
+            // Error
+            int code = error["code"].get_int();
+            if (code == RPC_IN_WARMUP && gArgs.GetBoolArg("-rpcwait", false)) {
+                throw CConnectionFailed("server in warmup");
```
After commit _cli: extract connection try/wait/failure logic_, this throw no longer gets caught.

jonatack replied:
Thanks @jnewbery, good catch on the throw! Updated that commit as per
`git diff e9a271b ed06899`
```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 53c6f3a655..18317574ad 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -419,6 +420,13 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler *rh, const std::string& str
    const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
     do {
         try {
             response = CallRPC(rh, strMethod, args, rpcwallet);
+
+            if (fWait) {
+                const UniValue& error = find_value(response, "error");
+                if (!error.isNull() && error["code"].get_int() == RPC_IN_WARMUP) {
+                    throw CConnectionFailed("server in warmup");
+                }
+            }
             break; // Connection succeeded, no need to retry.
         }
         catch (const CConnectionFailed&) {
@@ -507,12 +514,8 @@ static int CommandLineRPC(int argc, char *argv[])
         const UniValue& error = find_value(reply, "error");
         if (!error.isNull()) {
             // Error
-            int code = error["code"].get_int();
-            if (code == RPC_IN_WARMUP && gArgs.GetBoolArg("-rpcwait", false)) {
-                throw CConnectionFailed("server in warmup");
-            }
             strPrint = "error: " + error.write();
-            nRet = abs(code);
+            nRet = abs(error["code"].get_int());
             if (error.isObject()) {
                 UniValue errCode = find_value(error, "code");
                 UniValue errMsg  = find_value(error, "message");
```

jnewbery commented:
Concept and approach ACK.

jonatack commented:
Lone travis failure is a build timeout.

jonatack commented:
Rebased, with improved tests.

promag commented on src/rpc/request.cpp:
```diff
@@ -130,20 +130,21 @@ void DeleteAuthCookie()
+    const size_t num {in.size()};
     std::vector<UniValue> batch(num);
-    for (size_t i=0; i<in.size(); ++i) {
+    for (size_t i=0; i<num; ++i) {
```
c2d63a8c05e14cdfdaae2ebca7a6b38907c31f1c

```diff
     const size_t num {in.size()};
     std::vector<UniValue> batch(num);
-    for (size_t i=0; i<num; ++i) {
-        const UniValue &rec = in[i];
+    for (const UniValue &rec : in.getValues()) {
         if (!rec.isObject()) {
             throw std::runtime_error("Batch member must be an object");
```

jonatack replied:
That's better -- thanks! Done.

promag commented on src/bitcoin-cli.cpp:
```diff
@@ -275,9 +273,6 @@ class GetinfoRequestHandler: public BaseRequestHandler
             }
             result.pushKV("paytxfee", batch[ID_WALLETINFO]["result"]["paytxfee"]);
         }
-        if (!batch[ID_BALANCES]["result"].isNull()) {
```
Any reason to drop this (and break existing scripts)? Especially when `-rpcwallet` is set - and in this case all balances aren't even necessary.

jonatack replied:
Thanks for reviewing @promag. Hm, I don't know if API stability with -getinfo is an issue if it's intended for human use. With just the default wallet loaded the PR currently displays:
```
{
  "version": 209900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "keypoolsize": 1000,
  "paytxfee": 0.00000000,
  "relayfee": 0.00001000,
  "warnings": "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications",
  "balances": {
    "": 0.00000000
  }
}
```
ISTM it's better to display the balances consistently, whether one wallet or several?

jonatack replied:
Maybe I'm used to it, but I like having the balances set off in braces... it's easier to see them.

promag replied:
I mean that you could leave "balance" untouched and just add `balances`. Later we could drop it. Only a suggestion, but
 1. if `-rpcwallet` is set why show all balances? Sounds conceptually wrong. Note that `getwalletinfo` uses that (or are you planning to call `getwalletinfo` for all wallets?
 2. this is not an API but it's a command people can use in scripts so why break it for no good reason?

jonatack replied:
Yes, `getwalletinfo` adds a "walletname" field when `-rpcwallet` is set, but -getinfo does not. (FWIW it's odd that some field names use underscores and others don't... I'm guessing the snakecased ones are newer). It's late but I'll think about this.

promag replied:
Note that if multiple wallets are loaded and if `-rpcwallet` is not set then `getwalletinfo` fails, which results in not displaying `keypoolsize` and `paytxfee`.

So I think if `-rpcwallet` is set, `balances` shouldn't be displayed.

jonatack replied:
>     2. this is not an API but it's a command people can use in scripts so why break it for no good reason?

Most of the proposed ideas in #17314 could break scripts. Not breaking scripts would mean we would have to add things without removing or replacing them. This would make -getinfo unnecessarily long, but keeping -getinfo as short as possible is also a goal I've seen stated by long-term contributors.

jonatack replied:
> Note that if multiple wallets are loaded and if `-rpcwallet` is not set then `getwalletinfo` fails, which results in not displaying `keypoolsize` and `paytxfee`.
> 
> So I think if `-rpcwallet` is set (or just one wallet is loaded), `balances` shouldn't be displayed.

Ok, I will try this out and likely update the PR with this.

jonatack replied:
Done

michaelfolkson commented:
ACK ea59b18387c4b26240e1f087c568408922b3c940

Built, ran tests, tested manually by creating new wallet, transferring funds to it and unloading. Light code review.

jonatack commented:
PR updated per @promag's review feedback to now display a names + balances for all loaded wallets only when more than one wallet is loaded and `-rpcwallet=` is not passed; otherwise, behavior is unchanged. @fanquake, this PR no longer reverts the changes in #18574.

@jnewbery, @promag, @michaelfolkson, can you please re-review?

jb55 commented:
Concept ACK, but I wonder why in 208be5a811eee4f4d4d336a9e47c2b36760621b5 this is hardcoded to `getinfo`? Couldn't this be generalized to any command with almost no extra work (return results grouped by wallet name), or am I missing something?

jonatack commented:
Thanks @jb55, I agree, but there isn't currently consensus on what the server-side API interface should be and if one should be done at all. I learned from the first PR that a client-side version embedded into `-getinfo` as proposed here will be easier to merge due to more clear consensus and because it isn't subject to API constraints. I see this as a solution that can be merged now, and maybe a server-side one might emerge (or not) later, and could be called from -getinfo as well to batch the calls to `getbalances`.

jonatack commented:
Thanks @jb55, I agree, but there isn't currently consensus on what the server-side API interface should be and if one should be done at all. I learned from the first PR that a client-side version embedded into `-getinfo` as proposed here will be easier to merge due to more clear consensus and because it isn't subject to API constraints. I see this as a solution that can be merged now, and maybe a server-side one might emerge (or not) later, and could be called from -getinfo as well to batch the calls to `getbalances`. Don't hesitate to ACK if you'd like to see it move forward!

jonatack commented:
Rebased on master for the CI fix.

brakmic commented:
ACK https://github.com/bitcoin/bitcoin/commit/9fa9882fbbf4324f36cbf1008642adf43436be53

Built, run and tested on macOS Catalina 10.15.4
Manual testing with several wallets successful:

```shell
./src/bitcoin-cli -regtest -getinfo
{
  "version": 209900,
  "blocks": 205,
  "headers": 205,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "relayfee": 0.00001000,
  "warnings": "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications",
  "balances": {
    "": 5000.00000000,
    "other": 250.00000000,
    "secret_wallet": 0.00000000
  }
}
```

Execution of `./test/functional/wallet_multiwallet.py --usecli` was successful.

jnewbery commented on src/bitcoin-cli.cpp:
```diff
@@ -485,56 +512,54 @@ static int CommandLineRPC(int argc, char *argv[])
+            // was called without -rpcwallet and more than one wallet is loaded.
+            if (gArgs.GetBoolArg("-getinfo", false) && gArgs.GetArgs("-rpcwallet").empty()) {
+                rh.reset(new DefaultRequestHandler());
+                const UniValue listwallets = ConnectAndCallRPC(rh.get(), "listwallets", args);
```
`args` is always going to be empty here (see L239). Can you change this to an empty vector? (same for `ConnectAndCallRPRC()` call below)

jonatack replied:
Done.

jnewbery commented on src/rpc/request.cpp:
```diff
@@ -130,20 +130,20 @@ void DeleteAuthCookie()
 }
 
-std::vector<UniValue> JSONRPCProcessBatchReply(const UniValue &in, size_t num)
+std::vector<UniValue> JSONRPCProcessBatchReply(const UniValue& in)
```
Nice tidy up!

jnewbery reviewed (commented):
Good stuff Jon. A few style nits, but otherwise looks great

jonatack committed:
cli: use GetWalletBalances() functionality for -getinfo

and replace GetBoolArg with IsArgSet as we only want
to know if the arg is passed; we do not need the value.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,35 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+            if (fWait) {
+                const UniValue& error = find_value(response, "error");
+                if (!error.isNull() && error["code"].get_int() == RPC_IN_WARMUP) {
+                    throw CConnectionFailed("server in warmup");
```
Thanks, John. I tried but it wasn't working out in a way that was tidier, so leaving it be for now.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -485,56 +512,54 @@ static int CommandLineRPC(int argc, char *argv[])
+                rh.reset(new DefaultRequestHandler());
+                const UniValue listwallets = ConnectAndCallRPC(rh.get(), "listwallets", args);
+                const UniValue& wallets = find_value(listwallets, "result");
+                    if (find_value(listwallets, "error").isNull() && wallets.size() > 1) {
```
Oops, thanks John! Fixed.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -369,14 +369,13 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
     std::string endpoint = "/";
-    if (!gArgs.GetArgs("-rpcwallet").empty()) {
-        std::string walletName = gArgs.GetArg("-rpcwallet", "");
+    if (rpcwallet || !gArgs.GetArgs("-rpcwallet").empty()) {
```
Great point. Done; much better.

jonatack commented:
Thanks everyone for the reviews and @jnewbery for the excellent suggestions -- done and also moved the `GetWalletBalances` code to its own function and added Doxygen documentation for it and `ConnectAndCallRPC`.

jnewbery reviewed (commented):
utACK 2e7d8b9bf584ab61295795f0521537fc3f1057df. I've left some more style nits inline, but feel free to ignore. I think this is ready for merge as-is.

There's no need to credit reviewers in the commit logs (if we did that everywhere, the commit logs would get very long!)

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,68 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+ *
+ * @param[in] rh         Pointer to RequestHandler.
+ * @param[in] strMethod  Reference to const string method to forward to CallRPC.
+ * @param[in] rpcwallet  Pointer to const c-string rpcwallet arg to forward to CallRPC.
```
Good catch! Fixed.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,68 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+ * GetWalletBalances calls listwallets; if more than one wallet is loaded, it then
+ * fetches mine.trusted balances for each loaded wallet and pushes them to `result`.
+ *
+ * @param[in] result  Reference to UniValue object the wallet names and balances are pushed to.
```
Thanks. Replaced with just `@param`.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,68 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+ */
+static void GetWalletBalances(UniValue& result)
+{
+    std::unique_ptr<BaseRequestHandler> rh;
+    rh.reset(new DefaultRequestHandler());
```
TIL, thank you! Updated.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,68 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+{
+    std::unique_ptr<BaseRequestHandler> rh;
+    rh.reset(new DefaultRequestHandler());
+    const std::vector<std::string> args;
+    const UniValue listwallets = ConnectAndCallRPC(rh.get(), "listwallets", args);
```
Done.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,68 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+    const UniValue listwallets = ConnectAndCallRPC(rh.get(), "listwallets", args);
+    const UniValue& wallets = find_value(listwallets, "result");
+
+    if (find_value(listwallets, "error").isNull() && wallets.size() > 1) {
```
I agree (fan of guard clauses myself) but the conditional becomes a bit less readable to my eye going from an and to a negative or, so separated it into two guard conditionals.

jnewbery replied:
I think it's fine like that, but this is really a matter of personal taste. You should go with whichever way you prefer.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -486,55 +546,45 @@ static int CommandLineRPC(int argc, char *argv[])
+            reply = ConnectAndCallRPC(rh.get(), method, args, gArgs.GetArg("-rpcwallet", "").c_str());
+        } else {
+            reply = ConnectAndCallRPC(rh.get(), method, args);
+        }
```
I wanted to do this as well, but with C-style strings it causes issues (unless wrapped in the same block scope IIRC) and `test/functional/wallet_multiwallet.py --usecli` fails.

jonatack replied:
Inlined the wallet_name logic as a ternary in one call.

jnewbery replied:
My suggestion has a horrible bug, which only revealed itself when the wallet name is long enough. `c_str()` returns a `char*` that's only valid as long as the underlying `std::string` is in scope. In `gArgs.GetArg("-rpcwallet", "").c_str()`, the `gArgs.GetArg("-rpcwallet", "")` is a temporary that goes out of scope as soon as the statement ends, so `wallet_name` is pointing to freed memory. For short wallet names, that memory might not get reused, and the code that dereferences it still works, but for longer names, it fails.

I'm sorry to keep suggesting you move this around, but how about using a `Optional<std::string>` as the optional wallet name? The `ConnectAndCallRPC()` function signatures becomes:

`static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, Optional<std::string> rpcwallet = {})`

the wallet parsing code becomes:

```
        Optional<std::string> wallet_name {};

        if (gArgs.IsArgSet("-rpcwallet")) wallet_name = gArgs.GetArg("-rpcwallet", "");
```

and the code in `CallRPC()` changes to:

```
    // check if we should use a special wallet endpoint
    std::string endpoint = "/";
    if (rpcwallet) {
        char* encodedURI = evhttp_uriencode(rpcwallet->data(), rpcwallet->size(), false);
...
```

jonatack replied:
Thanks, John! Much better. Suggestions gratefully taken, passed as const reference (per https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in), added the explicit `#include <optional.h>` header, and changed `GetWalletBalances` to pass `const std::string wallet_name`.

jonatack replied:
(I had run into this `c_str()` issue 3 weeks ago and wondered what the better solution was. Thanks again!)

jonatack commented:
Thank you @jnewbery, I appreciate your outstanding reviewing and applied your suggestions. I also applied Clang formatting to the PR changeset. Here are the changes since the last push:

<details><summary><code>git diff 2e7d8b9 0af7d3a</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index a3302835c3..feea064be9 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -305,7 +305,7 @@ public:
     }
 };
 
-static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, const std::vector<std::string>& args, const char * const rpcwallet = nullptr)
+static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const char* const rpcwallet = nullptr)
 {
     std::string host;
     // In preference order, we choose the following for the port:
@@ -371,9 +371,9 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
     // check if we should use a special wallet endpoint
     std::string endpoint = "/";
     if (rpcwallet) {
-        char *encodedURI = evhttp_uriencode(rpcwallet, strlen(rpcwallet), false);
+        char* encodedURI = evhttp_uriencode(rpcwallet, strlen(rpcwallet), false);
         if (encodedURI) {
-            endpoint = "/wallet/"+ std::string(encodedURI);
+            endpoint = "/wallet/" + std::string(encodedURI);
             free(encodedURI);
         } else {
             throw CConnectionFailed("uri-encode failed");
@@ -426,7 +426,7 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
  * @returns the RPC response as a UniValue object.
  * @throws a CConnectionFailed std::runtime_error if connection failed or RPC server still in warmup.
  */
-static UniValue ConnectAndCallRPC(BaseRequestHandler *rh, const std::string& strMethod, const std::vector<std::string>& args, const char * const rpcwallet = nullptr)
+static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const char* const rpcwallet = nullptr)
 {
     UniValue response(UniValue::VOBJ);
     // Execute and handle connection failures with -rpcwait.
@@ -441,8 +441,7 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler *rh, const std::string& str
                 }
             }
             break; // Connection succeeded, no need to retry.
-        }
-        catch (const CConnectionFailed&) {
+        } catch (const CConnectionFailed&) {
             if (fWait) {
                 UninterruptibleSleep(std::chrono::milliseconds{1000});
             } else {
@@ -457,26 +456,24 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler *rh, const std::string& str
  * GetWalletBalances calls listwallets; if more than one wallet is loaded, it then
  * fetches mine.trusted balances for each loaded wallet and pushes them to `result`.
  *
- * @param[in] result  Reference to UniValue object the wallet names and balances are pushed to.
+ * @param result  Reference to UniValue object the wallet names and balances are pushed to.
  */
 static void GetWalletBalances(UniValue& result)
 {
-    std::unique_ptr<BaseRequestHandler> rh;
-    rh.reset(new DefaultRequestHandler());
-    const std::vector<std::string> args;
-    const UniValue listwallets = ConnectAndCallRPC(rh.get(), "listwallets", args);
+    std::unique_ptr<BaseRequestHandler> rh{MakeUnique<DefaultRequestHandler>()};
+    const UniValue listwallets = ConnectAndCallRPC(rh.get(), "listwallets", /* args=*/{});
+    if (!find_value(listwallets, "error").isNull()) return;
     const UniValue& wallets = find_value(listwallets, "result");
+    if (wallets.size() <= 1) return;

-    if (find_value(listwallets, "error").isNull() && wallets.size() > 1) {
-        UniValue balances(UniValue::VOBJ);
-        for (const UniValue& wallet : wallets.getValues()) {
-            const char * const wallet_name = wallet.get_str().c_str();
-            const UniValue getbalances = ConnectAndCallRPC(rh.get(), "getbalances", args, wallet_name);
-            const UniValue& balance = find_value(getbalances, "result")["mine"]["trusted"];
-            balances.pushKV(wallet_name, balance);
-        }
-        result.pushKV("balances", balances);

+    UniValue balances(UniValue::VOBJ);
+    for (const UniValue& wallet : wallets.getValues()) {
+        const char* const wallet_name = wallet.get_str().c_str();
+        const UniValue getbalances = ConnectAndCallRPC(rh.get(), "getbalances", /* args=*/{}, wallet_name);
+        const UniValue& balance = find_value(getbalances, "result")["mine"]["trusted"];
+        balances.pushKV(wallet_name, balance);
     }
+    result.pushKV("balances", balances);
 }
 
 static int CommandLineRPC(int argc, char *argv[])
@@ -545,13 +542,7 @@ static int CommandLineRPC(int argc, char *argv[])
             method = args[0];
             args.erase(args.begin()); // Remove trailing method name from arguments vector
         }
-
-        UniValue reply;
-        if (gArgs.IsArgSet("-rpcwallet")) {
-            reply = ConnectAndCallRPC(rh.get(), method, args, gArgs.GetArg("-rpcwallet", "").c_str());
-        } else {
-            reply = ConnectAndCallRPC(rh.get(), method, args);
-        }
+        const UniValue reply = ConnectAndCallRPC(rh.get(), method, args, gArgs.IsArgSet("-rpcwallet") ? gArgs.GetArg("-rpcwallet", "").c_str() : nullptr);
 
@@ -585,12 +576,10 @@ static int CommandLineRPC(int argc, char *argv[])
                 strPrint = result.write(2);
             }
         }
-    }
-    catch (const std::exception& e) {
+    } catch (const std::exception& e) {
         strPrint = std::string("error: ") + e.what();
         nRet = EXIT_FAILURE;
-    }
-    catch (...) {
+    } catch (...) {
         PrintExceptionContinue(nullptr, "CommandLineRPC()");
         throw;
     }
```
</p></details>

jnewbery commented:
@jonatack thanks for being so responsive to review, and well done for catching the nasty bug in my previous suggestion! (https://github.com/bitcoin/bitcoin/pull/18594#discussion_r428327325).

I have one final suggestion: switch out the c-style `char *` arguments for `Optional<std::string>`s. Using `std::string` seems better in almost all cases.

luke-jr commented on src/bitcoin-cli.cpp:
```diff
@@ -474,9 +532,8 @@ static int CommandLineRPC(int argc, char *argv[])
         std::unique_ptr<BaseRequestHandler> rh;
         std::string method;
-        if (gArgs.GetBoolArg("-getinfo", false)) {
+        if (gArgs.IsArgSet("-getinfo")) {
```
Why are you changing this?

jonatack replied:
`IsArgSet()` seemed more appropriate than `GetBoolArg()` as we don't need the value, only to know if the arg is set (and `IsArgSet()` is used elsewhere in this file for the same purpose). Updated the commit message to say why.

luke-jr replied:
Pretty sure this will do the wrong thing with `-nogetinfo` or `-getinfo=0` ?

jonatack replied:
Thanks -- with your examples and `IsArgSet`, it runs as if `-getinfo` was passed. With `GetBoolArg`, it raises with "error: too few parameters (need at least command)". If that is the desired behavior, then ISTM I should not only revert this change but also use `GetBoolArg` at line 569 and update the tests to cover this. Confirm?

jonatack replied:
Am also checking for `if (!gArgs.GetArgs("-rpcwallet").empty())` versus `gArgs.IsArgSet("-rpcwallet")`

jonatack replied:
Added a regression test since this was failing silently and reverted to parsing the `-getinfo` and `-rpcwallet` args strictly as before. Thanks @luke-jr for the catch!

```diff
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -532,7 +532,7 @@ static int CommandLineRPC(int argc, char *argv[])
         }
         std::unique_ptr<BaseRequestHandler> rh;
         std::string method;
-        if (gArgs.IsArgSet("-getinfo")) {
+        if (gArgs.GetBoolArg("-getinfo", false)) {
             rh.reset(new GetinfoRequestHandler());
@@ -543,7 +543,7 @@ static int CommandLineRPC(int argc, char *argv[])
             args.erase(args.begin()); // Remove trailing method name from arguments vector
         }
         Optional<std::string> wallet_name{};
-        if (gArgs.IsArgSet("-rpcwallet")) wallet_name = gArgs.GetArg("-rpcwallet", "");
+        if (!gArgs.GetArgs("-rpcwallet").empty()) wallet_name = gArgs.GetArg("-rpcwallet", "");
         const UniValue reply = ConnectAndCallRPC(rh.get(), method, args, wallet_name);
@@ -566,7 +566,7 @@ static int CommandLineRPC(int argc, char *argv[])
         } else {
-            if (gArgs.IsArgSet("-getinfo") && !gArgs.IsArgSet("-rpcwallet")) {
+            if (gArgs.GetBoolArg("-getinfo", false) && gArgs.GetArgs("-rpcwallet").empty()) {
                 GetWalletBalances(result); // fetch multiwallet balances and append to result
             }
diff --git a/test/functional/interface_bitcoin_cli.py b/test/functional/interface_bitcoin_cli.py
index 7530e7daf6..10ac82b107 100755
--- a/test/functional/interface_bitcoin_cli.py
+++ b/test/functional/interface_bitcoin_cli.py
@@ -49,6 +49,10 @@ class TestBitcoinCli(BitcoinTestFramework):
         self.log.info("Test -getinfo with arguments fails")
         assert_raises_process_error(1, "-getinfo takes no arguments", self.nodes[0].cli('-getinfo').help)
 
+        self.log.info("Test -getinfo=0 and -nogetinfo fail")
+        for command in ['-getinfo=0', '-nogetinfo']:
+            assert_raises_process_error(1, "error: too few parameters (need at least command)", self.nodes[0].cli(command).send_cli)
+
         self.log.info("Test -getinfo returns expected network and blockchain info")
```

jnewbery replied:
Seems fine, although this is a pretty pathological case. Why would anyone ever call `bitcoin-cli -nogetinfo`?

jonatack replied:
Fix and test coverage added in #19089

luke-jr commented on src/bitcoin-cli.cpp:
```diff
@@ -418,6 +417,65 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
+
+    UniValue balances(UniValue::VOBJ);
+    for (const UniValue& wallet : wallets.getValues()) {
+        const char* const wallet_name = wallet.get_str().c_str();
```
Scoping issue. `wallet_name` will be an out-of-scope temporary here.

Suggest just passing it as an arg, and using the UniValue in pushKV below.

```C++
    for (const UniValue& wallet : wallets.getValues()) {
        const UniValue getbalances = ConnectAndCallRPC(rh.get(), "getbalances", /* args=*/{}, wallet.get_str().c_str());
        const UniValue& balance = find_value(getbalances, "result")["mine"]["trusted"];
        balances.pushKV(wallet, balance);
    }
```

jonatack replied:
Good catch. Switched to std::string after taking @jnewbery's suggestion to switch to using Optional<std::string> for the `rpcwallet` parameter.

jonatack replied:
Good catch, thanks! Switched to `std::string` after taking @jnewbery's suggestion to switch to using `Optional<std::string>` for the `rpcwallet` parameter.

jonatack commented:
Thanks @jnewbery and @luke-jr for your reviews and suggestions. I guess it's not for nothing that the developer notes warn about `c_str()`. Updated as per the following diff.

<details><summary><code>git diff 0af7d3a 5edad5c</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
 #include <clientversion.h>
+#include <optional.h>
 #include <rpc/client.h>
@@ -17,7 +18,6 @@
 #include <util/url.h>
 
-#include <cstring>
 #include <functional>
@@ -305,7 +305,7 @@ public:
     }
 };
 
-static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const char* const rpcwallet = nullptr)
+static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const Optional<std::string>& rpcwallet = {})
 {
     std::string host;
     // In preference order, we choose the following for the port:
@@ -371,7 +371,7 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
     // check if we should use a special wallet endpoint
     std::string endpoint = "/";
     if (rpcwallet) {
-        char* encodedURI = evhttp_uriencode(rpcwallet, strlen(rpcwallet), false);
+        char* encodedURI = evhttp_uriencode(rpcwallet->data(), rpcwallet->size(), false);
         if (encodedURI) {
             endpoint = "/wallet/" + std::string(encodedURI);
             free(encodedURI);
@@ -422,11 +422,11 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
  *
  * @param[in] rh         Pointer to RequestHandler.
  * @param[in] strMethod  Reference to const string method to forward to CallRPC.
- * @param[in] rpcwallet  Pointer to const c-string rpcwallet arg to forward to CallRPC.
+ * @param[in] rpcwallet  Reference to const optional string wallet name to forward to CallRPC.
  * @returns the RPC response as a UniValue object.
  * @throws a CConnectionFailed std::runtime_error if connection failed or RPC server still in warmup.
  */
-static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const char* const rpcwallet = nullptr)
+static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const Optional<std::string>& rpcwallet = {})
 {
     UniValue response(UniValue::VOBJ);
     // Execute and handle connection failures with -rpcwait.
@@ -468,7 +468,7 @@ static void GetWalletBalances(UniValue& result)
 
     UniValue balances(UniValue::VOBJ);
     for (const UniValue& wallet : wallets.getValues()) {
-        const char* const wallet_name = wallet.get_str().c_str();
+        const std::string wallet_name = wallet.get_str();
         const UniValue getbalances = ConnectAndCallRPC(rh.get(), "getbalances", /* args=*/{}, wallet_name);
         const UniValue& balance = find_value(getbalances, "result")["mine"]["trusted"];
         balances.pushKV(wallet_name, balance);
@@ -542,7 +542,9 @@ static int CommandLineRPC(int argc, char *argv[])
             method = args[0];
             args.erase(args.begin()); // Remove trailing method name from arguments vector
         }
-        const UniValue reply = ConnectAndCallRPC(rh.get(), method, args, gArgs.IsArgSet("-rpcwallet") ? gArgs.GetArg("-rpcwallet", "").c_str() : nullptr);
+        Optional<std::string> wallet_name{};
+        if (gArgs.IsArgSet("-rpcwallet")) wallet_name = gArgs.GetArg("-rpcwallet", "");
+        const UniValue reply = ConnectAndCallRPC(rh.get(), method, args, wallet_name);
```
</p></details>

jonatack commented:
Thank you @jnewbery and @luke-jr for your excellent reviews and suggestions. I'll take more heed henceforth of the developer notes' warning about `c_str()`. Updated to use `Optional<std::string>` as per the following diff. Aside from a release note if merged, this PR should hopefully be ready.

<details><summary><code>git diff 0af7d3a 5edad5c</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
 #include <clientversion.h>
+#include <optional.h>
 #include <rpc/client.h>
@@ -17,7 +18,6 @@
 #include <util/url.h>
 
-#include <cstring>
 #include <functional>
@@ -305,7 +305,7 @@ public:
     }
 };
 
-static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const char* const rpcwallet = nullptr)
+static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const Optional<std::string>& rpcwallet = {})
 {
     std::string host;
     // In preference order, we choose the following for the port:
@@ -371,7 +371,7 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
     // check if we should use a special wallet endpoint
     std::string endpoint = "/";
     if (rpcwallet) {
-        char* encodedURI = evhttp_uriencode(rpcwallet, strlen(rpcwallet), false);
+        char* encodedURI = evhttp_uriencode(rpcwallet->data(), rpcwallet->size(), false);
         if (encodedURI) {
             endpoint = "/wallet/" + std::string(encodedURI);
             free(encodedURI);
@@ -422,11 +422,11 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
  *
  * @param[in] rh         Pointer to RequestHandler.
  * @param[in] strMethod  Reference to const string method to forward to CallRPC.
- * @param[in] rpcwallet  Pointer to const c-string rpcwallet arg to forward to CallRPC.
+ * @param[in] rpcwallet  Reference to const optional string wallet name to forward to CallRPC.
  * @returns the RPC response as a UniValue object.
  * @throws a CConnectionFailed std::runtime_error if connection failed or RPC server still in warmup.
  */
-static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const char* const rpcwallet = nullptr)
+static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& strMethod, const std::vector<std::string>& args, const Optional<std::string>& rpcwallet = {})
 {
     UniValue response(UniValue::VOBJ);
     // Execute and handle connection failures with -rpcwait.
@@ -468,7 +468,7 @@ static void GetWalletBalances(UniValue& result)
 
     UniValue balances(UniValue::VOBJ);
     for (const UniValue& wallet : wallets.getValues()) {
-        const char* const wallet_name = wallet.get_str().c_str();
+        const std::string wallet_name = wallet.get_str();
         const UniValue getbalances = ConnectAndCallRPC(rh.get(), "getbalances", /* args=*/{}, wallet_name);
         const UniValue& balance = find_value(getbalances, "result")["mine"]["trusted"];
         balances.pushKV(wallet_name, balance);
@@ -542,7 +542,9 @@ static int CommandLineRPC(int argc, char *argv[])
             method = args[0];
             args.erase(args.begin()); // Remove trailing method name from arguments vector
         }
-        const UniValue reply = ConnectAndCallRPC(rh.get(), method, args, gArgs.IsArgSet("-rpcwallet") ? gArgs.GetArg("-rpcwallet", "").c_str() : nullptr);
+        Optional<std::string> wallet_name{};
+        if (gArgs.IsArgSet("-rpcwallet")) wallet_name = gArgs.GetArg("-rpcwallet", "");
+        const UniValue reply = ConnectAndCallRPC(rh.get(), method, args, wallet_name);
```
</p></details>

promag commented:
Tested ACK 5edad5ce5d3f15b694bf3fad0300c6446674b554.

Does it make sense sum up balances and always have the `balance`?

jonatack commented:
@promag If this is merged, there are some separate follow-ups I could imagine: a release note, an attempt at reviving #17700 with an updated implementation (since it also does sequential calls like this PR), and sure, could propose a `total_balance` (`balance` might be misleading?) added after in `GetWalletBalances()`, with updated tests:
```json
  "balances": {
    "": 0.00001000,
    "Encrypted": 0.00003500,
    "day-to-day": 0.00000120,
    "side project": 0.00000094
  }
  "total_balance": ....
}
```
What I miss with -getinfo in single-wallet mode (as a user, not an RPC software client) is showing the wallet name. At some point a common interface like the above for both modes might be good, to always see the wallet name even when there is only one.

jnewbery commented:
utACK 5edad5ce5d3f15b694bf3fad0300c6446674b554

Thanks for being patient with my iterative reviews!

jonatack commented:
Thanks @jnewbery, @promag and @luke-jr for reviewing. I'll add a commit to the release note follow-up that tightens up the -getinfo command parsing and adds a test as per https://github.com/bitcoin/bitcoin/pull/18594#discussion_r428790740.

instagibbs commented:
concept ACK

meshcollider reviewed (commented):
Code review ACK 5edad5ce5d3f15b694bf3fad0300c6446674b554

Agree that a release note followup which can include a few other things would be good, this is RTM as is.

jonatack commented:
> Agree that a release note followup which can include a few other things would be good

Thanks -- done in #19089.
